internal/autoevent/executor.go

```diff
@@ -0,0 +1,94 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+//
+// Copyright (C) 2019 IOTech Ltd
+//
+// SPDX-License-Identifier: Apache-2.0
+
+package autoevent
+
+import (
+	"fmt"
+	"time"
+
+	"github.com/edgexfoundry/device-sdk-go/internal/common"
+	"github.com/edgexfoundry/device-sdk-go/internal/handler"
+	"github.com/edgexfoundry/go-mod-core-contracts/models"
+)
+
+const millionSecPerDay uint32 = 86400000
+
+type Executor interface {
+	Run()
+	Stop()
+}
+
+type executor struct {
+	deviceName   string
+	autoEvent    models.AutoEvent
+	lastReadings map[string]string
+	duration     time.Duration
+	stop         bool
+}
+
+func (e *executor) Run() {
+	for {
+		if e.stop {
+			break
+		}
+		time.Sleep(e.duration)
+
+		common.LoggingClient.Debug(fmt.Sprintf("AutoEvent - executing %v", e.autoEvent))
+		evt, appErr := readResource(e)
+		if appErr != nil {
+			common.LoggingClient.Error(fmt.Sprintf("AutoEvent - error occurs when read resource %s",
+				e.autoEvent.Resource))
+			continue
+		} else if evt != nil {
+			if e.autoEvent.OnChange {
+				if compareReadings(e, evt.Readings) {
+					common.LoggingClient.Debug(fmt.Sprintf("AutoEvent - readings are the same as previous one %v", e.lastReadings))
+					continue
+				}
+				cacheReadings(e, evt.Readings)
+			}
+			common.LoggingClient.Debug(fmt.Sprintf("AutoEvent - pushing event %s", evt.String()))
+			go common.SendEvent(evt)
+		}
+	}
+}
+
+func readResource(e *executor) (*models.Event, common.AppError) {
+	vars := make(map[string]string, 2)
+	vars[common.NameVar] = e.deviceName
+	vars[common.CommandVar] = e.autoEvent.Resource
+
+	evt, appErr := handler.CommandHandler(vars, "", common.GetCmdMethod)
+	return evt, appErr
+}
+
+func cacheReadings(e *executor, readings []models.Reading) {
+	for _, r := range readings {
+		e.lastReadings[r.Name] = r.Value
+	}
+}
+
+func compareReadings(e *executor, readings []models.Reading) bool {
+	for _, r := range readings {
+		v, ok := e.lastReadings[r.Name]
+		if !ok || v != r.Value {
+			return false
+		}
+	}
+	return true
+}
+
+func (e *executor) Stop() {
+	e.stop = true
+}
+
+func NewExecutor(deviceName string, ae models.AutoEvent) Executor {
```

Comment:
Shouldn't internal code still have godoc for public functions?

Reply:
ok, will add